segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeperstorage/LogStorageManager.java

```diff
@@ -0,0 +1,817 @@
+                                                                      logStorage.setUpdateVersion(logStorage.getUpdateVersion() + 1);
+                                                                      /** Fence all the ledgers and add one at the end. */
+                                                                      return fenceLedgersAndCreateOneAtEnd(segmentName, logStorage);
+                                                                  })
+                                                                  .thenApply(logStorage -> null);
+                                                      }))
+                   /** Delete metadata and cache for the source once the operation is complete. */
+                   .thenCompose(v -> this.zkClient.delete()
+                                                  .withOptions(new HashSet<>(
+                                                          Arrays.asList(DeleteOption.guaranteed,
+                                                                  DeleteOption.deletingChildrenIfNeeded)))
+                                                  .forPath(getZkPath(sourceSegment))
+                                                  .toCompletableFuture()
+                                                  .thenApply(ret -> {
+                                                      ledgers.remove(sourceSegment);
+                                                      return null;
+                                                  }));
+    }
+
+    /**
+     * Deletes a segment along with all the data and metadata.
+     * @param segmentName name of the segment to be deleted.
+     * @return A CompletableFuture which completes once the delete is complete.
+     */
+    public CompletableFuture<Void> delete(String segmentName) {
+        return this.getOrRetrieveStorageLedger(segmentName, false)
+                   .thenCompose(ledger -> {
+                       if (ledger.getContainerEpoch() > this.containerEpoch) {
+                           throw new CompletionException(new StorageNotPrimaryException(segmentName));
+                       }
+                       return this.zkClient.delete()
+                                           .withOptions(new HashSet<>(
+                                                   Arrays.asList(DeleteOption.guaranteed,
+                                                           DeleteOption.deletingChildrenIfNeeded)))
+                                           .forPath(getZkPath(segmentName))
+                                           .toCompletableFuture()
+                                           .thenApply(v -> {
+                                               ledgers.remove(segmentName);
+                                               return ledger;
+                                           });
+                   })
+                   .thenCompose(ledger -> ledger.deleteAllLedgers());
```

Comment:
You can move this up 3 lines where you have the `thenApply` so you can simplify your code.

Reply:
Fixed.